### PR TITLE
Bug 935060 - [Messages] Unable to send empty message

### PR DESCRIPTION
**Pull request by kumarrishav**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=71b3c0dd2d8c837173de9c3471a666b9f694560e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e8a534ad889ae19e79cda6449920156c0ffae658

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=1b633298f8d2ae06312441f05988e5ffd38ac854

**Comment by julienw** [comment]
wondering if you forgot to add your changes to `shared/js/multi_sim_action_button.js` ?

File: apps/sms/views/conversation/js/conversation.js [hunk 1]
```diff
@@ -688,7 +688,7 @@ var ConversationView = {
       // Populate draft if there is one
       // TODO merge with handleDraft ? Bug 1164431
       Drafts.request().then(() => {
-        this.draft = Threads.get(threadId).getDraft();
+        this.draft = Threads.get(threadId) && Threads.get(threadId).getDraft();
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: __question:__ why do we need this change?

File: apps/sms/views/conversation/js/conversation.js [hunk 2]
```diff
@@ -2248,18 +2244,25 @@ var ConversationView = {
...
+  simSelectedCallback: function conv_simSelected(cardIndex) {
+    var content = Compose.getContent();
     cardIndex = +cardIndex;
+
     if (isNaN(cardIndex)) {
       cardIndex = 0;
     }
 
-    this.sendMessage({ serviceId: cardIndex });
+    if(Compose.type == 'sms' && !content[0]) {
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: __question:__ Mmm, why do we do this only for ```sms```? Why can't we just check ```Compose.isEmpty()```?

File: apps/sms/views/conversation/js/conversation.js (hunk 1 above)
Comment (kumarrishav) [inline apps/sms/views/conversation/js/conversation.js]: @azasypkin  umm, it was throwing type error, i.e it's null, for some unit test (not related to this patch, this issue was before only). So, thought to fix it.

File: apps/sms/views/conversation/js/conversation.js (hunk 2 above)
Comment (kumarrishav) [inline apps/sms/views/conversation/js/conversation.js]: hmm, yup, Compose.isEmpty() should work. Will look into it.

File: apps/sms/views/conversation/js/conversation.js (hunk 1 above)
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: I think it's better to figure out real reason ;) We can do it later if we have the same issue in master

File: apps/sms/views/conversation/js/compose.js [hunk 3]
```diff
@@ -630,10 +630,9 @@ var Compose = (function() {
       var sizeExceedsLimit = this.type === 'mms' &&
         Settings.mmsSizeLimitation && this.size > Settings.mmsSizeLimitation;
 
-      // We can't send if message is empty, we're in the process of resizing or
+      // We can't send if we're in the process of resizing or
       // message is already too big.
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/compose.js]: __nit:__ since we have now more space at the top comment line, please move more words there from the bottom one :)

File: apps/sms/views/conversation/test/unit/compose_test.js [hunk 4]
```diff
@@ -1127,16 +1127,20 @@ suite('compose_test.js', function() {
         assert.isTrue(sendButton.disabled);
       });
 
-      test('enabled if both lock and Compose allow sending', function() {
```
Comment (azasypkin) [inline apps/sms/views/conversation/test/unit/compose_test.js]: __nit:__ I think the old title is still valid, since Compose has it's own check (e.g. size check).

File: apps/sms/views/conversation/test/unit/compose_test.js [hunk 5]
```diff
@@ -1127,16 +1127,20 @@ suite('compose_test.js', function() {
...
         Compose.refresh();
 
-        assert.isTrue(sendButton.disabled);
+        assert.isFalse(sendButton.disabled);
 
         Compose.append('Hola');
         assert.isFalse(sendButton.disabled);
       });
 
+      test('enabled when there is no message input', function() {
```
Comment (azasypkin) [inline apps/sms/views/conversation/test/unit/compose_test.js]: __nit:__ Honestly I don't see value in both this test and the test below (```enabled when there is message input```) - it's already verified in the test above - so I think we can safely remove these two tests.

File: apps/sms/views/conversation/js/conversation.js [hunk 6]
```diff
@@ -2248,18 +2244,25 @@ var ConversationView = {
     this.container.classList.remove('hide');
   },
 
-  // FIXME/bug 983411: phoneNumber not needed.
-  simSelectedCallback: function conv_simSelected(phoneNumber, cardIndex) {
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: __issue:__ why you removed ```phoneNumber``` parameter?

File: apps/sms/views/conversation/test/unit/conversation_test.js [hunk 7]
```diff
@@ -5110,14 +5110,10 @@ suite('conversation.js >', function() {
...
   suite('Sending Behavior (onSendClick)', function() {
-    function clickButton() {
-      clickButtonAndSelectSim(0);
-    }
-
-    function clickButtonAndSelectSim(serviceId) {
+    function clickButtonAndSelectSim(serviceId = 0) {
       ConversationView.onSendClick();
       // we get a string from the MultiSimActionButton
-      ConversationView.simSelectedCallback(undefined, '' + serviceId);
```
Comment (azasypkin) [inline apps/sms/views/conversation/test/unit/conversation_test.js]: related to my question about ```phoneNumber``` above, so you'll likely have to return this ```undefined``` here

File: apps/sms/views/conversation/test/unit/conversation_integration_test.js [hunk 8]
```diff
@@ -405,7 +405,7 @@ suite('ConversationView Integration', function() {
 
       // Simulate sendButton click
       ConversationView.onSendClick();
-      ConversationView.simSelectedCallback(undefined, 0);
```
Comment (azasypkin) [inline apps/sms/views/conversation/test/unit/conversation_integration_test.js]: same here

File: apps/sms/views/conversation/js/conversation.js (hunk 6 above)
Comment (kumarrishav) [inline apps/sms/views/conversation/js/conversation.js]: :o Nope, I was trying to fix bug 983411 here. as phoneNumber was redundant. PhoneNumber argument is of no use, i guess. Anyway i will fix it in that bug @azasypkin

Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: Ha-ha, yeah, but you forgot about changes for multisim action button here :) Anyway it's better to fix it in a dedicated bug and keep PR as small as possible (not like I usually do :P - I'm working on this :))

File: apps/sms/views/conversation/test/unit/compose_test.js (hunk 4 above)
Comment (kumarrishav) [inline apps/sms/views/conversation/test/unit/compose_test.js]: umm, i removed this because, now compose always allow sending. There is no check on compose now, as we can send empty message also.

File: apps/sms/views/conversation/test/unit/compose_test.js (hunk 5 above)
Comment (kumarrishav) [inline apps/sms/views/conversation/test/unit/compose_test.js]: okay :)

File: apps/sms/views/conversation/test/unit/conversation_test.js (hunk 7 above)
Comment (kumarrishav) [inline apps/sms/views/conversation/test/unit/conversation_test.js]: @azasypkin  That's why i removed phoneNumber argument as it was not used anywhere in code.

File: apps/sms/views/conversation/js/conversation.js (hunk 6 above)
Comment (kumarrishav) [inline apps/sms/views/conversation/js/conversation.js]: ha ha . sure :)  Julienw reminded me about multisim action thing. Will fix in other dedicated bug.

File: apps/sms/views/conversation/test/unit/conversation_test.js (hunk 7 above)
Comment (azasypkin) [inline apps/sms/views/conversation/test/unit/conversation_test.js]: Yeah, as I mentioned above - MultiSimActionButton still passes phoneNumber as first argument :)

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=80c9bef2321a9c53f944ed3da8a9405e704edf35

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9d1b80ba55aa8fb065c6115cacd12e4a6d8fad85

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=00327df99ee77d4276896ee4e71188add8f60b08

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8d4f17d774953dcfaa6669ffb34816de3cd71f0b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a71b7dee8b0a543f5c66377d7b1c459831d2ffe3

File: apps/sms/views/conversation/js/conversation.js [hunk 9]
```diff
@@ -2262,16 +2258,23 @@ var ConversationView = {
...
-
     cardIndex = +cardIndex;
+
     if (isNaN(cardIndex)) {
       cardIndex = 0;
     }
 
-    this.sendMessage({ serviceId: cardIndex });
+    if(Compose.isEmpty()) {
+      return Utils.confirm(
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: __nit:__ how about smth like this to have single entry point for ```sendMessage```?

```
    var sendConfirmationPromise = Compose.isEmpty() ?
      Utils.confirm(
        'sendEmptyMessage-confirmation', null, { className: 'danger' }
      ) :
      Promise.resolve();

    return sendConfirmationPromise.then(
      () => this.sendMessage({ serviceId: cardIndex })
    );
```

It will require minor changes in existing tests, but should not be too much

File: apps/sms/views/conversation/test/unit/conversation_test.js [hunk 10]
```diff
@@ -5252,10 +5248,39 @@ suite('conversation.js >', function() {
...
+    test('send empty SMS to one Recipient', function() {
+      var recipient = '999';
+
+      ConversationView.recipients.add({
+        number: recipient
+      });
+
+      Compose.getContent.returns([]); //return empty message
+
+      clickButtonAndSelectSim();
```
Comment (azasypkin) [inline apps/sms/views/conversation/test/unit/conversation_test.js]: With my recommendation, you'll need to have ```clickButtonAndSelectSim().then(() => ..... asserts....).then(done, done)``` everywhere it's used currently :)

Comment (azasypkin) [inline apps/sms/views/conversation/test/unit/conversation_test.js]: Also please add one more test for the case when user __doesn't__ confirm sending of empty message

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=92321903a243055bfc5aaa45b142ce7ad8867301

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ec031c8fca987dac3918129ed6e6d358c037e0bb

File: apps/sms/views/conversation/js/conversation.js [hunk 11]
```diff
@@ -2262,16 +2258,23 @@ var ConversationView = {
...
-    this.sendMessage({ serviceId: cardIndex });
+    var sendConfirmationPromise = Compose.isEmpty() ?
+      Utils.confirm(
+        'sendEmptyMessage-confirmation', null, { className: 'danger' }
+      ) :
+      Promise.resolve();
+
+    return sendConfirmationPromise.then(
+      () => this.sendMessage({ serviceId: cardIndex }), () => {
+        console.log('user doesn\'t confirm sending of empty messages');
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: __nit:__ I wouldn't do logging here, it doesn't seem useful, moreover with this ```simSelectedCallback``` will always return ```resolved``` promise. Yes it will require an extra changes in unit tests (see below), but it's more logically correct IMO.

So since ```Promise.reject()``` of ```Utils.confirm``` doesn't contain any error we should do an extra check in unit tests
```
// When we expect promise to be resolved
clickButtonAndSelectSim().then(() => {
  ... asserts...
}, () => {
  throw new Error('Promise should not be rejected!');
}).then(done, done);
.....
// Otherwise
clickButtonAndSelectSim().then(() => {
  throw new Error('Promise should not be resolved!');
}, () => {
  ... asserts...
}).then(done, done);
```

File: apps/sms/views/conversation/js/conversation.js [hunk 12]
```diff
@@ -2262,16 +2258,23 @@ var ConversationView = {
 
   // FIXME/bug 983411: phoneNumber not needed.
   simSelectedCallback: function conv_simSelected(phoneNumber, cardIndex) {
-    if (Compose.isEmpty()) {
-      return;
-    }
-
     cardIndex = +cardIndex;
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: __nit:__ actually you need to manipulate with cardIndex only if user confirms sending, so it'd be reasonable to do like this. What do you think?

```
var sendConfirmationPromise = Compose.isEmpty() ?
  Utils.confirm(
    'sendEmptyMessage-confirmation', null, { className: 'danger' }
  ) :
  Promise.resolve();

return sendConfirmationPromise.then(() => {
  cardIndex = +cardIndex;

  if (isNaN(cardIndex)) {
    cardIndex = 0;
  }

  this.sendMessage({ serviceId: cardIndex });
});
```

Comment (kumarrishav) [inline apps/sms/views/conversation/js/conversation.js]: Yeah, you are right. will do

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f59a952be1239eeaddad06805a7a2f09ca0167de

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d396f3789b6378576f7c7053cc929648a2fc06fe

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e0de0f0eb3cfaee646d2b0219f6883c51751561f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b85f928b54bc147963dfc052ed963f8be4da165c
